### PR TITLE
Set whitelist domains

### PR DESCRIPTION
**Pull request by juliemar**
### Description
to run in production is important to define a list of valid domains. This PR there is a proposal to define a list of valid email domains to execute login. 

### How to test?
Set a new Environment Variable `WHITELIST_DOMAINS` with a list of valid domains.

```
WHITELIST_DOMAINS: ["@mydomai1.com.br","@mydomai2.com] 
```

### Expected behavior
in the login page, only users with a valid email will enter in the office. the users with invalid email wil appear a message `invalid email for this office` like the image:
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -122,8 +122,11 @@ ROOMS_DATA=[
 ```
 
 # Running in Production
-If you will run in production we strongly recomend to you close your environmen using an internal VPN. In this solution everybody with the link and a valid google credential can enter in your virtual office. Because of this is important to you mantain your environment closed. Or you can help us implementing to define a whitelist domains with environment variable. 
+If you will run in production we strongly recomend to you close your environmen using an internal VPN. In this solution everybody with the link and a valid google credential can enter in your virtual office. Because of this is important to you mantain your environment closed. Or you can define a whitelist environmet variable with the list of valid domains to enter in your office.
 
+```
+WHITELIST_DOMAINS: ["@mydomai1.com.br","@mydomai2.com]
```
Comment (alfakini) [inline README.md]:
```suggestion
WHITELIST_DOMAINS: ["@mydomai1.com.br", "@mydomai2.com"] 
```

File: backend/app/app.config.js [hunk 2]
```diff
@@ -8,4 +8,5 @@ module.exports = {
   CUSTOM_STYLE: process.env.CUSTOM_STYLE || "",
   HOST: "0.0.0.0",
   PORT: process.env.PORT || 8080,
+  WHITELIST_DOMAINS: process.env.WHITELIST_DOMAINS || "",
```
Comment (alfakini) [inline backend/app/app.config.js]:
```suggestion
  WHITELIST_DOMAINS: process.env.WHITELIST_DOMAINS || [],
```

To keep the same data structure I recommend the use of an empty list as default.

File: backend/app/controllers/white.list.domain.controller.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+class WhiteListDomainController {
+  constructor(whiteListDomains) {
+    if(whiteListDomains!=""){
```
Comment (alfakini) [inline backend/app/controllers/white.list.domain.controller.js]:
```suggestion
    if (whiteListDomains.length > 0) {
```

File: backend/app/controllers/white.list.domain.controller.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    	this.whitelistDomains = JSON.parse(whiteListDomains);
+    }else{
+    	this.whitelistDomains = "";
```
Comment (alfakini) [inline backend/app/controllers/white.list.domain.controller.js]:
```suggestion
    } else {
    	this.whitelistDomains = [];
```

As there is a default, do we really need the `else` here?

File: backend/app/controllers/white.list.domain.controller.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    }
+  }
+
+  isValidEmailInWhiteList(email){
+    if(this.whitelistDomains==""){
```
Comment (alfakini) [inline backend/app/controllers/white.list.domain.controller.js]:
```suggestion
    if (whiteListDomains.length == 0) {
```

**Comment by alesshh** [comment]
Hey @juliemar 
I think the responsibility to validade domains should be in a middleware (https://expressjs.com/en/guide/using-middleware.html)

Moreover, it doesn’t solve the problem 100%, an attacker can bind the email to the socket request, and bypass validation.

I think the backend should receive an auth credentials and should validate user against google auth API. What you think about it?

**Comment by juliemar** [comment]
Hi @alesshh . 

@megatroom is working to put the domains validation in a middleware. I'm closing this PR and wait for the @megatroom solution.

**Comment by arthurbailao** [comment]
@juliemar we really need a list of domains? The OAuth consent screen can restrict the access only to internal users that belongs to organization. Take a look:

image

The login of users from the outside fails with this error:
image

**Comment by alesshh** [comment]
@arthurbailao I like this suggestion, but it doesn't allow multiple domains per instance. Resultados Digitais has an outsourcing company that can't access the instance only with org access.

And, it does not eliminate the server validation.

**Comment by arthurbailao** [comment]
@ale

> @arthurbailao I like this suggestion, but it doesn't allow multiple domains per instance. Resultados Digitais has an outsourcing company that can't access the instance only with org access.
> 
> And, it does not eliminate the server validation.

Sure, makes a lot of sense!

**Pull request closed without merging**